File: Data Structures & Algorithms/reverse-a-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head == None:
            return None
            
        current = head.next
        rev_head = head
        rev_head.next = None

        while current!=None:
            print(current.val)
            temp = current
            current = current.next
            temp.next = rev_head
            rev_head = temp
        return rev_head

File: Data Structures & Algorithms/reorder-linked-list/submission-1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

/*
get N first 
*/
class Solution {
    public void reorderList(ListNode head) {
        if(head.next!=null){
            ListNode slow = head;
            ListNode fast = head.next;
            while(fast!=null && fast.next!=null){
                slow=slow.next;
                fast=fast.next.next;
            }
            ListNode holder = slow.next;
            ListNode previous = null;
            slow.next=null;
            slow=holder.next;
            holder.next=null;
            while(slow!=null){
                previous = slow;
                slow=slow.next;
                previous.next=holder;
                holder=previous;
            }

            boolean flip = true;
            ListNode original = head;
            ListNode standard = head.next;
            while(standard!=null || holder!=null){
                if(flip){
                    original.next = holder;
                    holder=holder.next;
                }else{
                    original.next = standard;
                    standard=standard.next;
                }
                original=original.next;
                flip=!flip;
            }
        }
    }
}
